multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +934,311 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+
+			hostName := address.TargetRef.Name
+			hostLocalName := rcsw.mirroredResourceName(hostName)
+			if _, found := localServiceNames[hostLocalName]; found {
+				continue
+			}
+			// use a shared informer
+			localService, err := rcsw.localAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(hostLocalName)
+			if err != nil {
+				if !kerrors.IsNotFound(err) {
+					return err
+				}
+
+				// otherwise, create the service
+				localService, err = rcsw.createHeadlessMirroredHost(ctx, address.TargetRef, remoteService, hostLocalName)
+				if err != nil {
+					return err
+				}
+			}
+
+			localServiceNames[hostLocalName] = struct{}{}
+			newAddresses = append(newAddresses, corev1.EndpointAddress{
+				Hostname: hostName,
+				IP:       localService.Spec.ClusterIP,
+			})
+
+		}
+
+		// copy ports, create subset
+		newSubsets = append(newSubsets, corev1.EndpointSubset{
+			Addresses: newAddresses,
+			Ports:     subset.DeepCopy().Ports,
+		})
+	}
+
+	matchLabels := map[string]string{
+		consts.MirroredRootHeadlessLabel: rcsw.mirroredResourceName(remoteService.Name),
+	}
+
+	services, err := rcsw.localAPIClient.Svc().Lister().List(labels.Set(matchLabels).AsSelector())
+	if err != nil {
+		return err
+	}
+
+	var errors []error
+	for _, service := range services {
+		if _, found := localServiceNames[service.Name]; found {
+			continue
+		}
+		err := rcsw.localAPIClient.Client.CoreV1().Services(service.Namespace).Delete(ctx, service.Name, metav1.DeleteOptions{})
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				errors = append(errors, fmt.Errorf("could not delete Service: %s/%s:%s", service.Namespace, service.Name, err))
+			}
+		}
+	}
+	if len(errors) > 0 {
+		return RetryableError{errors}
+	}
+
+	// Update
+	localEp.Subsets = newSubsets
+	_, err = rcsw.localAPIClient.Client.CoreV1().Endpoints(localEp.Namespace).Update(ctx, localEp, metav1.UpdateOptions{})
+	if err != nil {
+		return RetryableError{[]error{err}}
+	}
+
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) handleHeadlessEndpointsCreated(ctx context.Context, remoteEp *corev1.Endpoints) error {
+	remoteService, err := rcsw.remoteAPIClient.Svc().Lister().Services(remoteEp.Namespace).Get(remoteEp.Name)
+	if err != nil {
+		return err
+	}
+	localServiceName := rcsw.mirroredResourceName(remoteService.Name)
+	serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
+
+	// To mirror a headless service, we need to create a clusterIP for each
+	// host in its Endpoints.
+	hosts := make(map[string]struct{})
+	subsetsToCreate := make([]corev1.EndpointSubset, 0, len(remoteEp.Subsets))
+	for _, subset := range remoteEp.Subsets {
+		newAddresses := make([]corev1.EndpointAddress, 0, len(subset.Addresses))
+		for _, addr := range subset.Addresses {
+			// Create annotations for host's service
+			if addr.TargetRef == nil {
+				// ignore addresses w/o a backing pod (if they exist)
+				continue
+			}
+
+			if _, found := hosts[addr.TargetRef.Name]; found {
```

Comment:
Going through this again a few days after I wrote it, I'm not convinced we should skip processing a host if it already exists, would really appreciate someone double checking me here. Here is why I think we shouldn't do this:

An Endpoints object has X subsets, where X !=0  and a subset is represented as a group of addresses with a common set of ports. The set of all endpoints in an Endpoints object is the union of all subsets. This means an address with multiple associated ports (that may come from multiple containers in the pod) will be in more than 1 subset. No address may be under NotReadyAddresses and ReadyAddresses at the same time in one subset.

This leads me to believe that if we go through a subset and process a host that is then found in the next subset, the current logic will skip adding it to the 2nd subset since it will already be present in the map. I don't think skipping here brings any benefit -- since we're using an informer there is also no performance cost from retrieving the same associated clusterIP service multiple times during this loop.

We do a similar thing when we create Endpoints too. Any thoughts on this?

Reply:
After some off GitHub discussion we decided it does make sense to process all hosts even if it exists in the map already.

This is because it is valid for two subsets to exist in an endpoints object where each subset contains the same `TargetRef.Name` but different ports. Let's say subset 1 has a ready endpoint for `foo.svc:8080` and subset 2 has a ready endpoint for `foo.svc:8081`. If we were to skip the second `foo.svc` because it was already added to `hosts` by the first subset, then we would skip adding `8081` as a port that is ready for that service.

Processing each subset's hostnames regardless of it being present in the map or not ensures that the source cluster's endpoints object is closer to the actual state of the target cluster's endpoints object.